test/metric_dimension/global_append_dimensions_test.go

```diff
@@ -0,0 +1,101 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+//go:build !windows
+
+package metric_dimension
+
+import (
+	"github.com/aws/amazon-cloudwatch-agent-test/test/metric"
+	"github.com/aws/amazon-cloudwatch-agent-test/test/metric/dimension"
+	"github.com/aws/amazon-cloudwatch-agent-test/test/status"
+	"github.com/aws/amazon-cloudwatch-agent-test/test/test_runner"
+	"log"
+)
+
+type GlobalAppendDimensionsTestRunner struct {
+	test_runner.BaseTestRunner
+}
+
+var _ test_runner.ITestRunner = (*GlobalAppendDimensionsTestRunner)(nil)
+
+func (t *GlobalAppendDimensionsTestRunner) Validate() status.TestGroupResult {
+	metricsToFetch := t.GetMeasuredMetrics()
+	testResults := make([]status.TestResult, len(metricsToFetch))
+	for i, metricName := range metricsToFetch {
+		testResults[i] = t.validateNoAppendDimensionMetric(metricName)
+	}
+
+	return status.TestGroupResult{
+		Name:        t.GetTestName(),
+		TestResults: testResults,
+	}
+}
+
+func (t *GlobalAppendDimensionsTestRunner) GetTestName() string {
+	return "GlobalAppendDimension"
+}
+
+func (t *GlobalAppendDimensionsTestRunner) GetAgentConfigFileName() string {
+	return "global_append_dimension.json"
+}
+
+func (t *GlobalAppendDimensionsTestRunner) GetMeasuredMetrics() []string {
+	return []string{"cpu_time_active"}
+}
+
+func (t *GlobalAppendDimensionsTestRunner) validateNoAppendDimensionMetric(metricName string) status.TestResult {
+	testResult := status.TestResult{
+		Name:   metricName,
+		Status: status.FAILED,
+	}
+
+	expDims, failed := t.DimensionFactory.GetDimensions([]dimension.Instruction{
+		{
+			Key:   "ImageId",
+			Value: dimension.UnknownDimensionValue(),
+		},
+		{
+			Key:   "InstanceId",
+			Value: dimension.UnknownDimensionValue(),
+		},
+		{
+			Key:   "InstanceType",
+			Value: dimension.UnknownDimensionValue(),
+		},
```

Comment:
Does this work? We're using append dimensions in the config so these should have actual values. When we query for an unknown dimension here, it's going to return nothing, isn't it?

Reply:
I checked the providers. As you can see here:
https://github.com/aws/amazon-cloudwatch-agent-test/blob/main/test/metric/dimension/instanceid_provider.go#L59-L62
It's only when the value is unknown that the provider (instanceID/imageId/InstanceType) will retrieve the actual value from imds and use it to query the metrics

Reply:
Ohh I guess I misunderstood how it works